src/tls/connection.c

```diff
@@ -736,13 +741,19 @@ verify_peer_certificate (gnutls_session_t session)
 #endif
         }
     }
-  else if (ret != GNUTLS_E_NO_CERTIFICATE_FOUND)
+  else if (ret == GNUTLS_E_NO_CERTIFICATE_FOUND)
+    {
+      debug (CONNECTION, "verify_peer_certificate: peer certificate not found");
+      /* that's ok, client certs are optional */
+      ret = GNUTLS_E_SUCCESS;
+    }
+  else
     {
       warnx ("Verifying TLS peer failed: %s", gnutls_strerror (ret));
       return ret;
     }
 
-  return GNUTLS_E_SUCCESS;
+  return ret;
```

Comment:
This change is extremely dubious.

The only way we hit the bottom of the function is if `ret == 0` && `status == 0` (after masking).  This line will therefore always return `0`.  I appreciate the additional debugging output, but I don't like how this one very specific case falls through, while the others are handled, unless it's very clear that this is the "success" case (which, before the reordering, had two possible fallthroughs).

My suggestion: add an explicit `GNUTLS_E_SUCCESS` return in the case that `ret == 0` and `status` is fine (removing the last possible fallthrough case), and then remove this `return` at the end, allowing the compiler to warn us if something changes above.  You might even change the direction of the check on `status`, allowing for a quick return, and unindent the error-handling block.  That's all a question of taste, though.

Reply:
I didn't like the original code here as any bug in the condition evaluation above would err on the side of success instead of failure. That was my main  reason to replace the `return GNUTLS_E_SUCCESS` (not the debug output), as any unhandled `ret != 0` would then be caught.

However, adding a `ret == 0` assertion or explicit check and letting the compiler complain is a good idea, thanks! 

I won't rework it right away though, as we should first have the discussion with the sssd devs about the CA validation API.